### ISSUE TITLE
First setup:upgrade fails on AddMollieShipmentIdAttribute patch

### ISSUE THREAD

**Issue by LucaGallinari** [mollie/magento2#822]
**Describe the bug**
**Sometimes**, during the first setup:upgrade of the module, the procedure fails while adding the `mollie_shipment_id` to the `sales_shipment` table:

```
Module 'Hyva_Checkout':
Module 'MarkShust_DisableTwoFactorAuth':
Module 'Mollie_Payment':
Enabling caches:
Current status:
layout: 1
block_html: 1
full_page: 1
Unable to apply data patch Mollie\Payment\Setup\Patch\Data\AddMollieShipmentIdAttribute for module Mollie_Payment. Original exception message: DDL statements are not allowed in transactions
  error  in magento2.php on line 302:
exit code 1 (General error)
```

Theoretically this should always fails as you [shouldn't be able to add a column via DataPatch](https://magento.stackexchange.com/questions/277432/error-ddl-statements-are-not-allowed-in-transactions-while-running-setupupgrad/277555#277555), but sometimes, especially on the second run of the setup:upgrade, it adds the column properly without errors.

Btw, it should be replaced with a proper definition of the column via [db_schema](https://developer.adobe.com/commerce/php/development/components/declarative-schema/configuration/) file.

**Used versions**
- *Magento Version number*: 2.4.4-p9
- *Open source/Enterprise/B2b*: open source
- *Mollie version number*: 2.40.1

**Comment by Frank-Magmodules** [comment]
Hi @LucaGallinari,

Thank you for opening this issue and providing the details. We’ve had the setup scripts integrated into our Continuous Integration workflow for quite some time now, and we’ve never encountered this problem. Could it be that there is a conflict with another module in your setup?

**Comment by Frank-Magmodules** [comment]
Thanks @JeroenBoersma , we will take a look into it!

**Comment by JeroenBoersma** [comment]
So, I can reproduce it time after time...

```shell
# install a clean Magento, nothing else and run setup:install
$ rm -rf generated
$ composer require mollie/magento2
$ bin/magento setup:upgrade --keep-generated
# ....
Module 'Mollie_Payment':                                                                                 
Unable to apply data patch Mollie\Payment\Setup\Patch\Data\AddMollieShipmentIdAttribute for module Mollie_Payment. Original exception message: DDL statements are not allowed in transactions                      
$ bin/magento setup:upgrade
```

So my last comment was incomplete, the `--keep-generated` messes things up! I know for sure that the flag was there when I commented earlier, but I didn't think it was part of the problem.

@LucaGallinari did you also use the `--keep-generated` flag? Just to find the common problem, as today I ran multiple scenarios but only this flag made the problem repeatable.

System settings:
- Magento 2.4.7-p3
- PHP 8.2.25 with zend opcache enabled
- MariaDB 10.6
- Redis 7.4.1
- Opensearch 2.18.0
- RabbitMQ 4.0.3

**Comment by Frank-Magmodules** [comment]
That’s helpful, @JeroenBoersma! We spent this morning trying to reproduce it, and now it seems we have a good lead on where and with what setup this might be happening. Thanks—I’ll follow up on this soon!

**Comment by LucaGallinari** [comment]
> @LucaGallinari did you also use the `--keep-generated` flag? Just to find the common problem, as today I ran multiple scenarios but only this flag made the problem repeatable.

It could be it! Locally i don't use that flag infact i never had this problem, instead on our CI/deploy we are using that flag and it failed on first "runs".

Btw, as i pointed out in the original post, the correct way to add a column is via a Declarative Schema and NOT via a DataPatch. We never run into this kind of problems by using declarative schema.

**Comment by Frank-Magmodules** [comment]
Hi @LucaGallinari,

Thank you once again for reporting this issue. It was quite tricky to reproduce, but with some valuable input from @JeroenBoersma, we managed to resolve it.

In addition to removing two redundant patches, we also had to refactor another patch, as it turned out the module’s `di.xml` changes weren’t being loaded at that stage yet.

We’re closing this issue now, but feel free to comment or reopen it if needed.
Thanks again!
